### PR TITLE
Nostr dm notifications

### PR DESCRIPTION
**Pull request by zappityzap**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Description

Adds notification provider for the nostr protocol: https://github.com/nostr-protocol/nostr

Notifications are sent as encrypted direct messages.

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)
1-notification-setup
2-nostr-notifications
3-test-succeed
4-test-multiple-partial
5-test-multiple-succeed
6-test-failed
7-error-invalid-nsec
8-error-invalid-npub

### CODE REVIEW

File: server/notification-providers/nostr.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                content: ciphertext,
+            };
+            event.id = getEventHash(event);
+            event.sig = signEvent(event, senderPrivateKey);
+            events.push(event);
+        }
+
+        // Publish events to each relay
+        const relays = notification.relays.split("\n");
+        const relayCount = relays.length;
```
Comment (CommanderStorm) [inline server/notification-providers/nostr.js:49]: please inline this
```suggestion
```

File: server/notification-providers/nostr.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                relay.close();
+            }
+        }
+
+        // Report success or failure
+        if (successfulRelays > 0) {
+            return `${successfulRelays}/${relayCount} relays connected.`;
+        } else {
+            throw Error("Failed to connect to any relays.");
+        }
```
Comment (CommanderStorm) [inline server/notification-providers/nostr.js:75]:
```suggestion
        if (successfulRelays > 0) {
            throw Error("Failed to connect to any relays.");
        }
        return `${successfulRelays}/${relays.length} relays connected.`;
```

File: server/notification-providers/nostr.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    name = "nostr";
+
+    async send(notification, msg, monitorJSON = null, heartbeatJSON = null) {
+
+        // All DMs should have same timestamp
+        const createdAt = Math.floor(Date.now() / 1000);
+
+        const senderPrivateKey = await this.getPrivateKey(notification.sender);
+        const senderPublicKey = getPublicKey(senderPrivateKey);
+        const recipientsPublicKeys = await this.getPublicKeys(notification.recipients);
```
Comment (CommanderStorm) [inline server/notification-providers/nostr.js:25]: Please move this block further down to where it is used

File: server/notification-providers/nostr.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    nip19
+} = require("nostr-tools");
+require("websocket-polyfill");
+
+class Nostr extends NotificationProvider {
+
+    name = "nostr";
+
+    async send(notification, msg, monitorJSON = null, heartbeatJSON = null) {
+
```
Comment (CommanderStorm) [inline server/notification-providers/nostr.js:19]:
```suggestion
    name = "nostr";

    async send(notification, msg, monitorJSON = null, heartbeatJSON = null) {
```

File: server/notification-providers/nostr.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
+const NotificationProvider = require("./notification-provider");
+global.crypto = require("crypto");
```
Comment (CommanderStorm) [inline server/notification-providers/nostr.js:2]: Why is this line included?
```suggestion
```

File: server/notification-providers/nostr.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+const {
+    relayInit,
+    getPublicKey,
+    getEventHash,
+    signEvent,
+    nip04,
+    nip19
+} = require("nostr-tools");
+require("websocket-polyfill");
```
Comment (CommanderStorm) [inline server/notification-providers/nostr.js:12]: Why is this included?
No other of the providers include this.

If you think this is needed, let's split this out of this PR.

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Overall this seems pretty dinished, despite being a draft.
Is this intentional?

I left some code style marks

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Overall this seems pretty dinished, despite being a draft.
Is this intentional?

I left some code style marks

File: server/notification-providers/nostr.js (hunk 2 above)
Comment (zappityzap) [reply server/notification-providers/nostr.js]: Thanks, that is simpler but it should now check for succesfulRelays == 0 instead of > 0.

File: server/notification-providers/nostr.js (hunk 6 above)
Comment (zappityzap) [reply server/notification-providers/nostr.js:12]: Without it, on Node 18 there is an error from the nostr-tools module. Node 19+ doesn't need it.
```
TypeError: Cannot read properties of undefined (reading 'readyState')
    at Object.close (/home/ansible/uptime-kuma/node_modules/nostr-tools/lib/nostr.cjs.js:566:14)
    at Nostr.send (/home/ansible/uptime-kuma/server/notification-providers/nostr.js:66:23)
    at async Socket.<anonymous> (/home/ansible/uptime-kuma/server/server.js:1209:27)
```

File: server/notification-providers/nostr.js (hunk 5 above)
Comment (zappityzap) [reply server/notification-providers/nostr.js:2]: Without it, on Node 18 there is an error from the nostr-tools module that crypto is not defined. Node 19+ doesn't need it.  There is a closed issue for nostr-tools about it: https://github.com/nbd-wtf/nostr-tools/issues/192

I see they have a better method for Node 18 compatibility now, but I'm not sure how to implement it:
```
// node.js 18 and earlier requires globalThis.crypto polyfill.
import { webcrypto } from 'node:crypto';
// @ts-ignore
if (!globalThis.crypto) globalThis.crypto = webcrypto;
```
Here is the error with Node 18:
```
ReferenceError: crypto is not defined
    at Object.encrypt (/home/ansible/uptime-kuma/node_modules/nostr-tools/lib/nostr.cjs.js:994:19)
    at Nostr.send (/home/ansible/uptime-kuma/server/notification-providers/nostr.js:34:44)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Socket.<anonymous> (/home/ansible/uptime-kuma/server/server.js:1209:27)
```

**Comment by zappityzap** [comment]
> Overall this seems pretty dinished, despite being a draft. Is this intentional?
> 
> I left some code style marks

Thank you for the review and feedback. I'm new to open source contributions and JavaScript. I wanted to learn more about the Nostr protocol so I implemented this for my own use. The project contribution guidelines said to mark it as a draft.

File: server/notification-providers/nostr.js (hunk 6 above)
Comment (CommanderStorm) [reply server/notification-providers/nostr.js:12]: Please include a comment that this is a `polyfill for node <=  18`

File: server/notification-providers/nostr.js (hunk 5 above)
Comment (CommanderStorm) [reply server/notification-providers/nostr.js:2]: Please include a comment that this is a `polyfill for node <=  18`

**Comment by CommanderStorm** [comment]
> The project contribution guidelines said to mark it as a draft.

See https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma
Notification Providers are handled differently from general feautre/code changes.
Empty draft PRs are only for discussing significant changes (as they carry significant reviewer effort+may risk breaking things) like #128/#2129/#2693

File: server/notification-providers/nostr.js (hunk 2 above)
Comment (CommanderStorm) [reply server/notification-providers/nostr.js]: True, I updated my comment

File: server/notification-providers/nostr.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                }
+            } catch (error) {
+                continue;
+            } finally {
+                relay.close();
+            }
+        }
+
+        // Report success or failure
+        if (successfulRelays == 0) {
```
Comment (CommanderStorm) [inline server/notification-providers/nostr.js:64]:
```suggestion
        if (successfulRelays === 0) {
```

File: src/components/notifications/Nostr.vue [hunk 8]
```diff
@@ -0,0 +1,24 @@
+<template>
+    <div class="mb-3">
+        <label for="nostr-relays" class="form-label">{{ $t("nostrRelays") }}<span style="color: red;"><sup>*</sup></span></label>
+        <textarea id="nostr-relays" v-model="$parent.notification.relays" class="form-control" :required=true placeholder="wss://127.0.0.1:7777/"></textarea>
```
Comment (CommanderStorm) [inline src/components/notifications/Nostr.vue:4]: See the linting warning by the github actions

**Comment by louislam** [comment]
I just realized that this pr is not working on arm v7, because there is no pre-built binary for the `bufferutil` dependency. So I reverted it.

Feel free to continue this by opening a new pr if you have found a solution.

https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma
> Easy to install for non-Docker users, no native build dependency is needed (for x86_64/armv7/arm64), no extra config, no extra effort required to get it running
